src/metalookup/features/addblock_based.py

```diff
@@ -1,9 +1,127 @@
-from core.extractor import ExtractionMethod, MetadataBase, ProbabilityDeterminationMethod
-
-
-@MetadataBase.with_key()
-class Advertisement(MetadataBase):
-    urls = [
+import asyncio
+import logging
+from concurrent.futures import Executor
+from logging import Logger
+from typing import Optional
+from unittest import mock
+
+import adblockparser
+from aiohttp import ClientSession
+
+from metalookup.app.models import Explanation, StarCase
+from metalookup.core.extractor import Extractor, download_tag_lists
+from metalookup.core.website_manager import WebsiteData
+from metalookup.lib.tools import runtime
+
+_FOUND_LIST_MATCHES = "Found list matches"
+_FOUND_NO_LIST_MATCHES = "Found no list matches"
+
+
+class AdBlockBasedExtractor(Extractor[set[str]]):
+    """
+    Base class for features to be extracted with `adblockparser` package.
+
+    The returned extra data corresponds to a sample of the blocked links.
+    The sample size can be configured with the limit member variable.
+    """
+
+    urls: list[str] = []  # needs to be provided by derived class
+
+    def __init__(self, logger: Optional[Logger] = None):
+        self.logger = logger or logging.getLogger(__name__)
+        self.adblock_parser_options = {
+            "script": True,
+            "image": True,
+            "stylesheet": True,
+            "object": True,
+            "xmlhttprequest": True,
+            "object-subrequest": True,
+            "subdocument": True,
+            "document": True,
+            "elemhide": True,
+            "other": True,
+            "background": True,
+            "xbl": True,
+            "ping": True,
+            "dtd": True,
+            "media": True,
+            "third-party": True,
+            "match-case": True,
+            "collapse": True,
+            "donottrack": True,
+            "websocket": True,
+            "domain": "",
+        }
+        self.rules: AdblockRules = None  # noqa will be initialized in async setup call
+        self.limit: Optional[int] = 2
+        """
+        Stop after finding a given amount of to be blocked links.
+        For websites with a lot of referenced content, this can significantly speed up the extraction process,
+        as once it is clear that there will be blocked content, there is no more need to check for more blocked content.
+        If set to none, then all links will be checked.
+        """
+
+    async def setup(self) -> None:
+        """Fetch tag lists from configured urls."""
+        assert self.urls != [], "Missing url specification of tag lists"
```

Comment:
I stay with my opinion on asserts, i.e., they are implicit, like described here: https://stackoverflow.com/questions/1569049/making-pythons-assert-throw-an-exception-that-i-choose

Reply:
yeah, it's not really about which exception is raised here.

Assertions  (also in other languages - which also do have exceptions or other methods of error handling) are still regularly used in some code bases to state expected invariants about the state of some part of the program. 

In this case: The list must never be empty - otherwise this is a bug, and the application should or cannot recover from it.

This it was separates it from exceptions - they serve to handle some exceptional state which may occur and needs to be recovered from if possible.

Still: changed to raise an exception